src/flows/context/notebooks.api.tsx

```diff
@@ -0,0 +1,115 @@
+import {
+  PatchNotebookParams,
+  patchNotebook,
+  PostNotebookParams,
+  postNotebook,
+  DeleteNotebookParams,
+  deleteNotebook,
+  getNotebooks,
+} from 'src/client/notebooksRoutes'
+import {notebookUpdateFail} from 'src/shared/copy/notifications'
+import {notify} from 'src/shared/actions/notifications'
+
+const DEFAULT_API_FLOW: PatchNotebookParams = {
+  id: '',
+  data: {},
+}
+let stagedFlow: PatchNotebookParams = DEFAULT_API_FLOW
+let reportDecayTimeout = null
+let reportMaxTimeout = null
+
+const REPORT_DECAY = 500 // number of miliseconds to wait after last event before sending
+const REPORT_MAX_WAIT = 5000 // max number of miliseconds to wait between sends
+
+export const pooledUpdateAPI = (flow: PatchNotebookParams) => {
+  stagedFlow = flow
+
+  if (!!reportDecayTimeout) {
+    clearTimeout(reportDecayTimeout)
+    reportDecayTimeout = null
+  }
+
+  if (!reportMaxTimeout) {
+    reportMaxTimeout = setTimeout(() => {
+      reportMaxTimeout = null
+
+      // flow already synced
+      if (stagedFlow === DEFAULT_API_FLOW) {
+        return
+      }
+
+      clearTimeout(reportDecayTimeout)
+      reportDecayTimeout = null
+      updateAPI(stagedFlow)
+
+      stagedFlow = DEFAULT_API_FLOW
+    }, REPORT_MAX_WAIT)
+  }
+
+  reportDecayTimeout = setTimeout(() => {
+    updateAPI(stagedFlow)
+
+    stagedFlow = DEFAULT_API_FLOW
+  }, REPORT_DECAY)
+}
+
+export const updateAPI = async (flow: PatchNotebookParams) => {
+  const res = await patchNotebook(flow)
+  if (res.status != 200) {
+    throw new Error(res.data.message)
+  }
+}
+
+export const createAPI = async (flow: PostNotebookParams) => {
+  const res = await postNotebook(flow)
+  if (res.status != 200) {
+    throw new Error(res.data.message)
+  }
+  return res.data.id
```

Comment:
can we get rid of all of this duplication and handholding once we have our first migration?

Reply:
yes!